1-Draft/RFC00XX-Module-Isolation.md

```diff
@@ -0,0 +1,251 @@
+  because the `System.Management.Automation.dll` is always loaded in the default ALC.
+- `Assembly.LoadFrom` is hardcoded to always load an assembly into the default ALC.
+  It also sets up an event handler to `AppDomain.AssemblyResolve` to handle loading additional assemblies from the same folder path
+  for future implicit assembly loading requests.
+
+In addition, we have two more forms of assembly loading:
+- For the dynamic assembly generated by `Add-Type` via `Compilation.Emit`
+  we load it by `AssemblyLoadContext.Default.LoadFromStream`.
+- For the dynamic assembly generated by PowerShell class using the `System.Reflection.Emit` APIs,
+  it's loaded into the inferred active ALC, which is the default ALC today.
+
+The assembly loading logic in PowerShell needs to be updated to make it aware of the correct ALC to use for the loading.
+Furthermore, we need to set the contextual ALC before running a module command and restore it afterwards,
+so that the explicit or implicit assembly loading triggered by the module command execution can be handled by the correct ALC.
+
+- `AssemblyLoadContext.CurrentContextualReflectionContext` should be set before invoking a module command,
+  and restored to the previous context after the invocation.
+- `AssemblyLoadContext.CurrentContextualReflectionContext` should be set before running a command or script block in a module context (`& $module <cmd-or-script>`),
+  and restored to the previous context after the invocation.
+
+There is risk if we completely depend on the `AssemblyLoadContext.CurrentContextualReflectionContext` to get the active ALC to work with,
+because that property can be changed by a user.
+If `CurrentContextualReflectionContext` is changed to an arbitrary ALC,
+we may run into weird type/assembly resolution issues.
+
+So, I propose to add the new internal property `ExecutionContext.ActiveReflectionContext`,
+which points to the currently active ALC that is supposed to be visible and usable in PowerShell.
+If its value is `null`, then the currently active ALC is the default ALC.
+
+Similarly, the setup and restore of this property should be done before and after command invocation:
+- `ExecutionContext.ActiveReflectionContext` is set before invoking a module command,
+  and restored to the previous context afterwards.
+- `ExecutionContext.ActiveReflectionContext` is set before running a command or script block in a module context (`& $module <cmd-or-script>`),
+  and restored to the previous context after the invocation.
+
+### Update Type Resolution Logic
+
+Now that we have custom ALC's for modules, the type resolving logic needs to be changed:
+
+1. We should first look for the type within the `ExecutionContext.ActiveReflectionContext`,
+and then revert back to the default ALC.
+1. When reverting back to the default ALC,
+we should ignore the assemblies that are loaded (and thus already searched) from the active ALC.
```

Comment:
I would caution that we should not ignore assemblies of a different version that may be present in both contexts.

Reply:
When we are within a module context where the `ActiveReflectionContext` is not null (the module has a custom ALC associated), we should respect the module's explicit choices of assemblies. If a version of an assembly is already loaded in that custom ALC, then we should only inspect types from that version of assembly but not one with a different version in the default ALC. Doing the latter will likely result in a type identity issue when using it in the module.

Reply:
Note that, when we are not running in a module context, the `ExecutionContext.ActiveReflectionContext` will always be null, meaning the active context is the default ALC.